File: ios/ReactNativeCameraKit/CKMockPreview.m
//
//  CKMockPreview.m
//  ReactNativeCameraKit
//
//  Created by Aaron Grider on 10/20/20.
//

#import "CKMockPreview.h"

@implementation CKMockPreview

- (id)initWithFrame:(CGRect) frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit {
    self.layer.cornerRadius  = 10.0f;
    self.layer.masksToBounds = YES;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    [self randomize];
}

- (void)randomize {
    self.layer.backgroundColor = [UIColor colorWithHue:drand48() saturation:1.0 brightness:1.0 alpha:1.0].CGColor;
    self.layer.sublayers = nil;
    
    for (int i = 0; i < 5; i++) {
        [self drawBalloon];
    }
}

- (void)drawBalloon {
    int stringLength = 200;
    CGFloat radius = [CKMockPreview randomNumberBetween:50 maxNumber:150];
    int x = arc4random_uniform(self.frame.size.width);
    int y = arc4random_uniform(self.frame.size.height + radius + stringLength);
    int stretch = radius / 3;
    
    CALayer *balloon = [CALayer layer];
    balloon.frame = CGRectMake(x - radius, y - radius, radius * 2, radius * 2 + stringLength);
    
    // Ballon main circle
    CAShapeLayer *circle = [CAShapeLayer layer];
    double colorHue = drand48();
    
    [circle setPath:[[UIBezierPath bezierPathWithOvalInRect:CGRectMake(0, 0, radius * 2, radius * 2 + stretch)] CGPath]];
    [circle setFillColor:[[UIColor colorWithHue:colorHue saturation:1.0 brightness:0.95 alpha:1.0] CGColor]];
    
    // Ballon reflection
    CAShapeLayer *reflection = [CAShapeLayer layer];
    [reflection setPath:[[UIBezierPath bezierPathWithOvalInRect:CGRectMake(radius / 2, radius / 2, radius * 0.7, radius * 0.7)] CGPath]];
    [reflection setFillColor:[[UIColor colorWithHue:colorHue saturation:1.0 brightness:1.0 alpha:1.0] CGColor]];
    
    // Ballon string
    CAShapeLayer *line = [CAShapeLayer layer];
    UIBezierPath *linePath= [UIBezierPath bezierPath];
    CGPoint startPoint = CGPointMake(balloon.frame.size.width / 2, radius * 2);
    CGPoint endPoint = CGPointMake(balloon.frame.size.width, (radius * 2) + stringLength);
    [linePath moveToPoint: startPoint];
    [linePath addQuadCurveToPoint:endPoint controlPoint:CGPointMake(balloon.frame.size.width / 2, radius * 2 + stringLength / 2)];
    line.path = linePath.CGPath;
    line.fillColor = nil;
    line.strokeColor = [UIColor darkGrayColor].CGColor;
    line.opacity = 1.0;
    line.lineWidth = radius * 0.05;
    
    // Add layers
    [balloon addSublayer:line];
    [circle addSublayer:reflection];
    [balloon addSublayer:circle];
    
    [self.layer addSublayer:balloon];
    
    // Apply animation
    CABasicAnimation *scale = [CABasicAnimation animationWithKeyPath:@"transform.scale"];
    [scale setFromValue:[NSNumber numberWithFloat:0.7f]];
    [scale setToValue:[NSNumber numberWithFloat:1.0f]];
    [scale setDuration:10.0f];
    [scale setFillMode:kCAFillModeForwards];
    
    scale.removedOnCompletion = NO;
    scale.autoreverses= YES;
    scale.repeatCount = HUGE_VALF;
    
    CABasicAnimation *move = [CABasicAnimation animationWithKeyPath:@"position.y"];
    [move setFromValue:[NSNumber numberWithFloat:balloon.frame.origin.y]];
    [move setToValue:[NSNumber numberWithFloat: 0 - balloon.frame.size.height]];
    [move setDuration:[CKMockPreview randomNumberBetween:30 maxNumber:100]];
    
    move.removedOnCompletion = NO;
    move.repeatCount = HUGE_VALF;

    [balloon addAnimation:scale forKey:@"scale"];
    [balloon addAnimation:move forKey:@"move"];
}


- (UIImage *)snapshotWithTimestamp:(BOOL)showTimestamp {
    UIGraphicsBeginImageContextWithOptions(self.bounds.size, NO, 0);
    [self drawViewHierarchyInRect:self.bounds afterScreenUpdates:NO];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    if (showTimestamp) {
        NSDate *date = [NSDate date];
        NSDateFormatter *dateformatter = [[NSDateFormatter alloc] init];
        [dateformatter setDateFormat:@"HH:mm:ss"];
        NSString *stringFromDate = [dateformatter stringFromDate:date];
        UIFont *font = [UIFont boldSystemFontOfSize:20];
        
        [image drawInRect:CGRectMake(0,0,image.size.width,image.size.height)];
        CGRect rect = CGRectMake(25, 25, image.size.width, image.size.height);
        [[UIColor whiteColor] set];
        [stringFromDate drawInRect:CGRectIntegral(rect) withAttributes:[NSDictionary dictionaryWithObject:font forKey:NSFontAttributeName]];
        
        image = UIGraphicsGetImageFromCurrentImageContext();
        
        UIGraphicsEndImageContext();
    }
    
    return image;
}

+ (NSInteger)randomNumberBetween:(NSInteger)min maxNumber:(NSInteger)max
{
    return min + arc4random_uniform((uint32_t)(max - min + 1));
}

@end

File: ios/ReactNativeCameraKit/CKMockPreview.swift
//
//  CKMockPreview.swift
//  ReactNativeCameraKit
//

import UIKit

@objc(CKMockPreview)
public class CKMockPreview: UIView {
    // MARK: - Public

    @objc
    public override init(frame: CGRect) {
        super.init(frame: frame)

        layer.cornerRadius = 10.0
        layer.masksToBounds = true
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override public func layoutSubviews() {
        super.layoutSubviews()

        randomize()
    }

    @objc
    public func snapshotWithTimestamp(_ showTimestamp: Bool) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(bounds.size, false, 0)
        drawHierarchy(in: bounds, afterScreenUpdates: false)
        var image = UIGraphicsGetImageFromCurrentImageContext()

        if showTimestamp {
            let date = Date()
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "HH:mm:ss"
            let stringFromDate = dateFormatter.string(from: date)
            let font = UIFont.boldSystemFont(ofSize: 20)

            image?.draw(in: CGRect(x: 0, y: 0, width: image?.size.width ?? 0, height: image?.size.height ?? 0))
            let rect = CGRect(x: 25, y: 25, width: image?.size.width ?? 0, height: image?.size.height ?? 0)
            UIColor.white.set()
            let textFontAttributes = [NSAttributedString.Key.font: font]
            stringFromDate.draw(in: rect.integral, withAttributes: textFontAttributes)

            image = UIGraphicsGetImageFromCurrentImageContext()
        }

        UIGraphicsEndImageContext()
        return image
    }

    @objc
    public func randomize() {
        layer.backgroundColor = UIColor(hue: CGFloat(Double.random(in: 0...1)), saturation: 1.0, brightness: 1.0, alpha: 1.0).cgColor
        layer.sublayers = nil

        for _ in 0..<5 {
            drawBalloon()
        }
    }

    // MARK: - Private

    private func drawBalloon() {
        let stringLength = CGFloat(200)
        let radius = CGFloat(Int.random(in: 50...150))

        let x = CGFloat(Int.random(in: 0...Int(frame.size.width)))
        let y = CGFloat(Int.random(in: 0...Int(frame.size.height + radius + stringLength)))
        let stretch = radius / 3

        let balloon = CALayer()
        balloon.frame = CGRect(x: x - radius, y: y - radius, width: radius * 2, height: radius * 2 + stringLength)

        // Balloon main circle
        let circle = CAShapeLayer()
        let colorHue = Double.random(in: 0...1)

        circle.path = UIBezierPath(ovalIn: CGRect(x: 0, y: 0, width: radius * 2, height: radius * 2 + stretch)).cgPath
        circle.fillColor = UIColor(hue: colorHue, saturation: 1.0, brightness: 0.95, alpha: 1.0).cgColor

        // Balloon reflection
        let reflection = CAShapeLayer()
        reflection.path = UIBezierPath(ovalIn: CGRect(x: radius / 2, y: radius / 2, width: radius * 0.7, height: radius * 0.7)).cgPath
        reflection.fillColor = UIColor(hue: colorHue, saturation: 1.0, brightness: 1.0, alpha: 1.0).cgColor

        // Balloon string
        let line = CAShapeLayer()
        let linePath = UIBezierPath()
        let startPoint = CGPoint(x: balloon.frame.size.width / 2, y: radius * 2)
        let endPoint = CGPoint(x: balloon.frame.size.width, y: (radius * 2) + stringLength)
        linePath.move(to: startPoint)
        linePath.addQuadCurve(to: endPoint, controlPoint: CGPoint(x: balloon.frame.size.width / 2, y: radius * 2 + stringLength / 2))
        line.path = linePath.cgPath
        line.fillColor = nil
        line.strokeColor = UIColor.darkGray.cgColor
        line.opacity = 1.0
        line.lineWidth = radius * 0.05

        // Add layers
        balloon.addSublayer(line)
        circle.addSublayer(reflection)
        balloon.addSublayer(circle)

        layer.addSublayer(balloon)

        // Apply animation
        let scale = CABasicAnimation(keyPath: "transform.scale")
        scale.fromValue = NSNumber(value: 0.7)
        scale.toValue = NSNumber(value: 1.0)
        scale.duration = 10.0
        scale.fillMode = .forwards
        scale.isRemovedOnCompletion = false
        scale.autoreverses = true
        scale.repeatCount = .greatestFiniteMagnitude

        let move = CABasicAnimation(keyPath: "position.y")
        move.fromValue = NSNumber(value: balloon.frame.origin.y)
        move.toValue = NSNumber(value: 0 - balloon.frame.size.height)
        move.duration = Double.random(in: 30...100)
        move.isRemovedOnCompletion = false
        move.repeatCount = .greatestFiniteMagnitude

        balloon.add(scale, forKey: "scale")
        balloon.add(move, forKey: "move")
    }
}
